core/src/main/scala-2.13+/cats/data/NonEmptyLazyList.scala

```diff
@@ -62,14 +64,70 @@ object NonEmptyLazyList extends NonEmptyLazyListInstances {
   def fromSeq[A](as: Seq[A]): Option[NonEmptyLazyList[A]] =
     if (as.nonEmpty) Option(create(LazyList.from(as))) else None
 
-  def fromLazyListPrepend[A](a: A, ca: LazyList[A]): NonEmptyLazyList[A] =
-    create(a +: ca)
+  def fromLazyListPrepend[A](a: => A, ll: => LazyList[A]): NonEmptyLazyList[A] =
+    create(a #:: ll)
+
+  @deprecated("Use overload with by-name args", "2.11.0")
+  def fromLazyListPrepend[A]()(a: A, ll: LazyList[A]): NonEmptyLazyList[A] =
+    fromLazyListPrepend(a, ll)
+
+  def fromLazyListAppend[A](ll: => LazyList[A], a: => A): NonEmptyLazyList[A] =
+    create(ll #::: a #:: LazyList.empty)
 
-  def fromLazyListAppend[A](ca: LazyList[A], a: A): NonEmptyLazyList[A] =
-    create(ca :+ a)
+  @deprecated("Use overload with by-name args", "2.11.0")
+  def fromLazyListAppend[A]()(ll: LazyList[A], a: A): NonEmptyLazyList[A] =
+    fromLazyListAppend(ll, a)
 
   def apply[A](a: => A, as: A*): NonEmptyLazyList[A] =
-    create(LazyList.concat(LazyList(a), LazyList.from(as)))
+    create(a #:: LazyList.from(as))
+
+  /**
+   * Wraps a `LazyList` that may or may not be empty so that individual
+   * elements or `NonEmptyLazyList`s can be prepended to it to construct a
+   * result that is guaranteed not to be empty without evaluating any elements.
+   */
+  def maybe[A](ll: => LazyList[A]): Maybe[A] = new Maybe(() => ll)
+
+  /**
+   * Wrapper around a `LazyList` that may or may not be empty so that individual
+   * elements or `NonEmptyLazyList`s can be prepended to it to construct a
+   * result that is guaranteed not to be empty without evaluating any elements.
+   */
+  final class Maybe[A] private[NonEmptyLazyList] (mkLL: () => LazyList[A]) {
+    // because instances of this class are created explicitly, they might be
+    // reused, and we don't want to re-evaluate `mkLL`
+    private[this] lazy val ll = mkLL()
+
+    /** Prepends a single element, yielding a `NonEmptyLazyList`. */
+    def #::[AA >: A](elem: => AA): NonEmptyLazyList[AA] =
+      create(elem #:: ll)
+
+    /** Prepends a `LazyList`, yielding another [[Maybe]]. */
+    def #:::[AA >: A](prefix: => LazyList[AA]): Maybe[AA] =
+      new Maybe(() => prefix #::: ll)
```

Comment:
I doubt this method belongs here.
Prepending one `LazyList` to another one lazily is the job that `LazyList.Deferrer` seems to be doing already. I.e.:
```scala
def a: LazyList[Int] = ???
def b: LazyList[Int] = ???

NonEmptyList.maybe(a #::: b)
// the above seems to be the same as
a #::: NonEmptyList.maybe(b)
```
Anyway a method that only involves `LazyList` on both sides does not seem to be a part of the `NonEmptyLazyList` DSL to me.

Reply:
so this question is... tough. conceivably, one might build up a result something like as follows:

```scala
def buildResult[A](maybe: NonEmptyLazyList.Maybe[A]): NonEmptyLazyList[A] =
  if (something) someNell #::: maybe
  else if (somethingElse) a #:: b #:: maybe
  else buildResult(someLL #::: maybe)
}
buildResult(NonEmptyLazyList.maybe(LazyList.empty))
```

now, you could argue that you still don't need it, and can construct the `Maybe` instance later. which is true. but that seems uglier and less intuitive to me. would love to get thoughts from more people

Reply:
>  you could argue that you still don't need it
Yeah, indeed, I would :)

Again just IMO: the entire `Maybe` does not seem that it should be a part of `NonEmptyLazyList` per se. Because it looks like a bunch of extension methods for `LazyList` that simply helps to construct a new `NonEmptyLazyList` out of it.

For example, see `cats.syntax.ListSyntax` and its corresponding `ListOps.toNel`, etc.

So perhaps it would make sense to consider creating a `cats.syntax.LazyListSyntax` which would provide all this functionality as a bunch of extension methods only.

Reply:
the issue is, `NonEmptyLazyList.fromLazyList` evaluates the first state/element of the `LazyList`, which is Bad™; `NonEmptyList.fromList` does not have this concern as the `List` is already fully evaluated. the intent behind `Maybe` was to have a way to construct a `NonEmptyLazyList` without evaluating any elements. now that `fromLazyListUnsafe` has been changed to be lazy, `Maybe` isn't needed _as_ much, though I maintain that it is still a useful tool. its design could perhaps use a bit more workshopping though

Reply:
I've been sick for a week, so I haven't been able to address things, but I'm hopefully better enough to get back to this.

the reason we need some sort of wrapper around `LazyList` with a `#:::` operator is that, despite it syntactically being a right-associative prepend operator, `#:::` actually behaves semantically as a left-associative append operator.

this also means that my example earlier was terrible, and the design of `Maybe` is pretty bad too. I really have been off my game here.

I'll re-evaluate and redesign this part of the API to be a bit more reasonable with this in mind